File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        stack = [root]
        res = []
        visited = dict()
        def downLeft(node):
            while(node.left and node.left not in visited):
                node = node.left
            return node
        while stack:
            top = stack[-1]
            leftmost=downLeft(top)
            if not leftmost.right or leftmost.right in visited:
                res.append(leftmost.val)
                visited[leftmost] = 1
                if leftmost == top:
                    stack.pop()
            else:
                if leftmost != top:
                    stack.append(leftmost)
                stack.append(leftmost.right)
        return res
                
                
            
        

File: 145-binary-tree-postorder-traversal/145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        LinkedList<TreeNode> stack = new LinkedList<>();
        TreeNode cur = root;
        TreeNode prev = null;
        while(cur != null || !stack.isEmpty()){
            while(cur != null){
                stack.addLast(cur);
                cur = cur.left;
            }
            
            TreeNode top = stack.peekLast();
            if(top.right == null || prev == top.right){
                prev = stack.pollLast();
                res.add(prev.val);
            }else{
                cur = top.right;
            }
        }
        return res;
    }
}